src/core/lib.lua

```diff
@@ -386,12 +353,64 @@ if ffi.abi("be") then
    function htonl(b) return b end
    function htons(b) return b end
 else
-   function htonl(b) return bit.bswap(b) end
-   function htons(b) return bit.rshift(bit.bswap(b), 16) end
+   function htonl(b) return bswap(b) end
+   function htons(b) return rshift(bswap(b), 16) end
 end
 ntohl = htonl
 ntohs = htons
 
+-- The fact that BitOps return signed integers is irritating, to say
+-- the least.  One counter-intuitive consequence is that, for example,
+--
+--   bit.bswap(bit.bswap(0x80000000)) == 0x80000000
+--
+-- is false, because the lhs is a negative number.  To compare a
+-- 32-bit quantity with the result of ntohl() or htonl(), one needs to
+-- use the equivalent of
+--
+--   bit.band(number, 0xFFFFFFFF) == ntohl(other_number)
+--
+-- The followig utility can be used for that purpose.
+function eq32 (a, b)
+   return band(a, 0xFFFFFFFF) == band(b, 0xFFFFFFFF)
+end
+
+-- Manipulation of bit fields in uint{8,16,32)_t stored in network
+-- byte order.  Using bit fields in C structs is compiler-dependent
+-- and a little awkward for handling endianness and fields that cross
+-- byte boundaries.  We're bound to the LuaJIT compiler, so I guess
+-- this would be save, but masking and shifting is guaranteed to be
+-- portable.
+
+local bitfield_endian_conversion = 
+   { [16] = { ntoh = ntohs, hton = htons },
+     [32] = { ntoh = ntohl, hton = htonl }
+  }
+
+function bitfield(size, struct, member, offset, nbits, value)
+   local conv = bitfield_endian_conversion[size]
```

Comment:
Yes, it was just moved after the definition of the endian-conversion stuff to avoid forward declarations.